front-end/src/game/GameScreen.tsx

```diff
@@ -1,90 +1,234 @@
+      let tileY = tiles[i]._y * this.state.zoom + this.state.offsetY + rect.y;
+      let tileW = tiles[i]._size * this.state.zoom;
+      let tileH = tiles[i]._size * this.state.zoom;
+
+      if (
+        mouseX > tileX &&
+        mouseX < tileX + tileW &&
+        mouseY > tileY &&
+        mouseY < tileY + tileH
+      ) {
+        console.log(tiles[i]._id);
+        console.log(tileW);
+        tiles[i]._selected = true;
+      } else {
+        tiles[i]._selected = false;
+      }
+    }
+  };
 
   mouseDrag = (e) => {
+    this.mouseHover(e);
     if (this.state.mouseDown) {
-      let oldMapX = this.state.mapX
-      let oldMapY = this.state.mapY
-      let oldMouseX = this.state.mouseX
-      let oldMouseY = this.state.mouseY
-  
-      let newMouseX = e.screenX
-      let newMouseY = e.screenY
-
-      let newMapX = oldMapX + newMouseX-oldMouseX
-      let newMapY = oldMapY + newMouseY-oldMouseY
-      
-      this.setState({mouseX: newMouseX, mouseY: newMouseY, mapX: newMapX, mapY: newMapY })
+      let oldMouseX = this.state.mouseX;
+      let oldMouseY = this.state.mouseY;
+
+      let newMouseX = e.clientX;
+      let newMouseY = e.clientY;
+
+      let newOffsetX = this.state.offsetX + newMouseX - oldMouseX;
+      let newOffsetY = this.state.offsetY + newMouseY - oldMouseY;
+
+      if (this.state.zoom === 1) {
+        newOffsetX = 0;
+        newOffsetY = 0;
+      }
+
+      this.setState({
+        mouseX: newMouseX,
+        mouseY: newMouseY,
+        offsetX: newOffsetX,
+        offsetY: newOffsetY
+      });
+    }
+  };
+
+  checkOffset = () => {
+    let newOffsetX = this.state.offsetX;
+    let newOffsetY = this.state.offsetY;
+    const rect = this.canvasRef.current.getBoundingClientRect();
+    if (
+      // right side
+      rect.x + rect.width >
+      rect.x + newOffsetX + this.state.mapWidth * this.state.zoom
+    ) {
+      newOffsetX = rect.width - this.state.mapWidth * this.state.zoom;
+    }
+
+    if (
+      // left side
+      rect.x <
+      rect.x + newOffsetX
+    ) {
+      newOffsetX = 0;
+    }
+
+    if (
+      // bottom side
+      rect.y + rect.height >
+      rect.y + newOffsetY + this.state.mapHeight * this.state.zoom
+    ) {
+      newOffsetY = rect.height - this.state.mapHeight * this.state.zoom;
+    }
+
+    if (
+      // top side
+      rect.y <
+      rect.y + newOffsetY
+    ) {
+      newOffsetY = 0;
     }
-  }
+
+    if (
+      newOffsetX !== this.state.offsetX ||
+      newOffsetY !== this.state.offsetY
+    ) {
+      this.setState({
+        offsetX: newOffsetX,
+        offsetY: newOffsetY
+      });
+    }
+  };
 
   mouseWheel = (e) => {
-    console.log(e)
-    let oldMapWidth = this.state.mapWidth
-    let oldMapHeight = this.state.mapHeight
-    let zoom = 1
-    
+    let zoom = this.state.zoom;
+
     if (e.deltaY < 0) {
-      zoom = 1.10
+      zoom += 0.1;
     } else {
-      zoom = 0.9
+      zoom -= 0.1;
     }
```

Comment:
mouseWheel = (e) => {
    let { zoom, offsetX, offsetY } = this.state;
  
    const rect = e.target.getBoundingClientRect();
    const mouseX = e.clientX - rect.left;
    const mouseY = e.clientY - rect.top;
  
    const zoomFactor = 1.1;
  
    let newZoom;
    if (e.deltaY < 0) {
      newZoom = zoom * zoomFactor;
    } else {
      newZoom = zoom / zoomFactor;
    }
  
    newZoom = Math.min(Math.max(newZoom, 1), 2);
  
    const zoomChange = newZoom / zoom;
  
    offsetX = mouseX - zoomChange * (mouseX - offsetX);
    offsetY = mouseY - zoomChange * (mouseY - offsetY);
  
    this.setState({
      zoom: newZoom,
      offsetX: offsetX,
      offsetY: offsetY,
    });
  
    e.preventDefault();
  };

Reply:
This code is aware of the position of the mouse and zooms accordingly.